core/ordering/cosipbft/blockstore/mem.go

```diff
@@ -258,20 +263,20 @@ func (obs *observer) pushAndWait() {
 
 func (obs *observer) close() {
 	obs.Lock()
-	defer obs.Unlock()
 
 	obs.closed = true
+	obs.running = false
+	obs.buffer = nil
 
-	if obs.running {
-		obs.running = false
-		obs.buffer = nil
-
-		// Drain the message in transit to close the channel properly.
-		select {
-		case <-obs.ch:
-		default:
-		}
+	// Drain message in transit to close the channel properly.
+	select {
+	case <-obs.ch:
+	default:
 	}
 
 	close(obs.ch)
```

Comment:
I may be wrong, but if there are multiple messages in the buffer, the `pushAndWait` will still try to send those messages, but the chan will be closed.
..

Ok, this works because `len(obs.buffer) == 0` will return true, I get it.